File: training/src/kingdoms_sim/character.cpp
#include "kingdoms_sim/character.hpp"

namespace kingdoms {

void Character::Tick(float dt_seconds) {
  attack_cooldown -= dt_seconds;
  if (attack_cooldown < -1.0f) attack_cooldown = -1.0f;

  health += regen * dt_seconds;
  if (health > max_health) health = max_health;
}

void Character::Damage(float amount) { health -= amount; }

void Character::ApplyGear(const AssetDatabase& assets) {
  const ItemData& helmet = assets.Item(head);
  const ItemData& torso = assets.Item(body);

  max_health = 25.0f * helmet.stats.health * torso.stats.health;
  regen = helmet.stats.regen * torso.stats.regen;
  speed = helmet.stats.speed * torso.stats.speed;
  power = helmet.stats.damage * torso.stats.damage;
  reload = helmet.stats.reload * torso.stats.reload;
}

bool Character::AddItemOrBust(uint8_t item_id) {
  for (uint8_t i = 0; i < kBackpackSlots; ++i) {
    if (inventory.find(i) == inventory.end()) {
      inventory[i] = item_id;
      return true;
    }
  }
  return false;
}

void Character::ChangeInventory(const AssetDatabase& assets, uint8_t to, uint8_t from) {
  if (to == from) return;

  auto is_gear = [](uint8_t slot) { return slot == kHeadSlot || slot == kBodySlot; };

  if (is_gear(to)) {
    uint8_t& to_gear = (to == kHeadSlot) ? head : body;
    ItemSlot to_type = (to == kHeadSlot) ? ItemSlot::Head : ItemSlot::Body;
    uint8_t to_default = (to == kHeadSlot) ? kDefaultHeadItem : kDefaultBodyItem;

    auto from_it = inventory.find(from);
    if (from_it == inventory.end() || assets.Item(from_it->second).slot != to_type) return;

    uint8_t from_item = from_it->second;
    uint8_t equipped = to_gear;
    to_gear = from_item;

    if (equipped == to_default) {
      inventory.erase(from);
    } else {
      inventory[from] = equipped;
    }
    ApplyGear(assets);
    return;
  }

  if (is_gear(from)) {
    uint8_t& from_gear = (from == kHeadSlot) ? head : body;
    ItemSlot from_type = (from == kHeadSlot) ? ItemSlot::Head : ItemSlot::Body;
    uint8_t from_default = (from == kHeadSlot) ? kDefaultHeadItem : kDefaultBodyItem;

    uint8_t equipped = from_gear;
    if (equipped == from_default) return;

    auto to_it = inventory.find(to);
    if (to_it != inventory.end()) {
      if (assets.Item(to_it->second).slot != from_type) return;
      from_gear = to_it->second;
    } else {
      from_gear = from_default;
    }
    inventory[to] = equipped;
    ApplyGear(assets);
    return;
  }

  auto from_it = inventory.find(from);
  if (from_it == inventory.end()) return;
  uint8_t item1 = from_it->second;

  auto to_it = inventory.find(to);
  if (to_it != inventory.end()) {
    inventory[from] = to_it->second;
  } else {
    inventory.erase(from);
  }
  inventory[to] = item1;
  ApplyGear(assets);
}

void Character::RemoveInventory(const AssetDatabase& assets, uint8_t slot) {
  if (slot < kBackpackSlots) {
    inventory.erase(slot);
  } else if (slot == kHeadSlot) {
    head = kDefaultHeadItem;
  } else {
    body = kDefaultBodyItem;
  }
  ApplyGear(assets);
}

Character MakeDefaultCharacter(const AssetDatabase& assets) {
  Character c;
  c.hand = 0;
  c.head = kDefaultHeadItem;
  c.body = kDefaultBodyItem;
  c.inventory[0] = 8;  // Kendo Stick; matches DefaultCharacter's starting weapon in character.go
  c.ApplyGear(assets);
  c.health = c.max_health;
  return c;
}

}  // namespace kingdoms

File: server/engine/character.go
package engine

import "kingdoms/engine/assets"

const (
	inventorySlots = 24
	headSlot       = 24
	bodySlot       = 25
	defaultHead    = 0
	defaultBody    = 1
)

type character struct {
	id             uint32
	position       Position
	angle          uint16
	hand           uint8
	head           uint8
	body           uint8
	maxHealth      float32
	health         float32
	regen          float32
	speed          float32
	power          float32
	reload         float32
	inventory      map[uint8]uint8
	attackCounter  uint8
	attackCooldown float32
	dead           bool
}

func newCharacter(id uint32, catalog *assets.Catalog) *character {
	character := &character{
		id:        id,
		head:      defaultHead,
		body:      defaultBody,
		inventory: map[uint8]uint8{0: 8},
	}
	character.apply(catalog)
	character.health = character.maxHealth
	return character
}

func (c *character) state() CharacterState {
	inventory := make(map[uint8]uint8, len(c.inventory))
	for slot, item := range c.inventory {
		inventory[slot] = item
	}

	return CharacterState{
		ID:        c.id,
		X:         c.position.X,
		Y:         c.position.Y,
		Angle:     c.angle,
		Health:    c.health,
		MaxHealth: c.maxHealth,
		Reload:    c.reload,
		Speed:     c.speed,
		Hand:      c.hand,
		Head:      c.head,
		Body:      c.body,
		Inventory: inventory,
	}
}

func (c *character) move(x, y float32, angle uint16) {
	c.position = Position{X: x, Y: y}
	c.angle = angle
}

func (c *character) tick(seconds float32) bool {
	previousHealth := uint16(c.health)
	c.attackCooldown = max(c.attackCooldown-seconds, -1)
	c.health = min(c.health+c.regen*seconds, c.maxHealth)
	return previousHealth != uint16(c.health)
}

func (c *character) damage(amount float32) bool {
	c.health = max(c.health-amount, 0)
	if c.health == 0 {
		c.dead = true
	}
	return c.dead
}

func (c *character) apply(catalog *assets.Catalog) {
	helmet, _ := catalog.Item(c.head)
	body, _ := catalog.Item(c.body)

	c.maxHealth = 25 *
		multiplier(helmet.Stats.Health) *
		multiplier(body.Stats.Health)
	c.regen = multiplier(helmet.Stats.Regen) *
		multiplier(body.Stats.Regen)
	c.speed = multiplier(helmet.Stats.Speed) *
		multiplier(body.Stats.Speed)
	c.power = multiplier(helmet.Stats.Damage) *
		multiplier(body.Stats.Damage)
	c.reload = multiplier(helmet.Stats.Reload) *
		multiplier(body.Stats.Reload)
	c.health = min(c.health, c.maxHealth)
}

func (c *character) addItem(item uint8) bool {
	for slot := uint8(0); slot < inventorySlots; slot++ {
		if _, occupied := c.inventory[slot]; !occupied {
			c.inventory[slot] = item
			return true
		}
	}
	return false
}
